### ISSUE TITLE
Review TOA outputs

### ISSUE THREAD

**Issue by virginiayung** [mapbox/rio-toa#15]
### tl;dr:

@celoyd noticed some weirdness in Greenland's before and after TOA histograms. We're double-checking the TOA process.
### How Greenland TOA was processed:

```
$ rio toa reflectance LC80270062016060LGN00_B4.TIF LC80270062016060LGN00_B3.TIF LC80270062016060LGN00_B2.TIF LC80270062016060LGN00_mtl.json output.TIF --dst-dtype uint16
```

From MTL:

```
M = [2e-05, 2e-05, 2e-05],
A = [-0.1, -0.1, -0.1],
E = 6.43073832,
sin(E)= 0.14701817733832565
```
### The behavior we expected to see:

The equation for calculating `reflectance` is as follows:

```
Reflectance = (M * img + A) / sin(E)
```

Rescaling:

```
Reflectance_arr *= rescale_factor * np.iinfo(np.uint16).max
np.clip(Reflectance_arr , 1, np.iinfo(np.uint16).max).astype(dtype)
```

So we expect to see a shift due to adding `AR` and a shape transformation from multiplying by`MR` and dividing by `sin(E)` in our post-TOA histogram. Like the _first 2 graphs_ @celoyd drew:

![histogram_parrots](https://cloud.githubusercontent.com/assets/4450007/17074478/7f8115e0-5030-11e6-8deb-9ef31996b5b7.gif)

![slack for ios upload 4](https://cloud.githubusercontent.com/assets/4450007/17074043/b6bea87e-502b-11e6-86a1-d93d5ed56543.jpg)
### This is what we're seeing:

We're seeing a very slight shift to the left because of `A = [-0.1, -0.1, -0.1]` and shape transformation to each of the RGB bands due to `M = [2e-05, 2e-05, 2e-05]` and `sin(E)= 0.14701817733832565`:
![image](https://cloud.githubusercontent.com/assets/4450007/17074026/778e320a-502b-11e6-8389-6efe31c8f1bd.png)
### Next Steps:

I'm writing some simple analysis tools in [ipython notebook](https://gist.github.com/virginiayung/5e991035cbce905db2e7e90407df87fa) to double-check if we're getting the correct outputs (aka not case #3 in @celoyd's graph☝️ ) and how the pixels get transformed by the TOA process.

cc @dnomadb @celoyd

**Comment by celoyd** [comment]
Also seeing something odd when I:
1. Match the histograms of a reflectance toa’d band to the original band,
2. take their difference, and
3. contrast-stretch that:

![w2](https://cloud.githubusercontent.com/assets/148985/17074651/b5d3b22c-5032-11e6-991e-34a060486193.jpeg)

Those bands are 512 px tall. I’m not sure this is a significant issue – it might be a tiny rounding problem or something. But worth tracking down.

Edit: to replicate this I’m using LC80460292016177LGN00 from the AWS PDS and running `rio toa reflectance -p --dst-dtype uint16 LC80460292016177LGN00/LC80460292016177LGN00_B3.TIF LC80460292016177LGN00/LC80460292016177LGN00_MTL.txt b3test.tif`, then matching and diffing `b3test.tif` with `LC80460292016177LGN00_B3.TIF`.

**Comment by virginiayung** [comment]
@celoyd I'm looking into [LaSRC toa product](https://github.com/USGS-EROS/espa-surface-reflectance/blob/master/not-validated-prototype-lasrc/c_version/src/l8sr_algorithm.txt), along with some logic dealing with the pan and thermal bands that we are not accounting for in rio-toa.  Referencing @perrygeo's ticket https://github.com/mapbox/rio-toa/issues/12. Here's a link to my [messy ipynb](https://gist.github.com/virginiayung/5e991035cbce905db2e7e90407df87fa)

**Comment by dnomadb** [comment]
@virginiayung @celoyd how do the results compare when doing per-pixel and not per-pixel sun elevation?

**Comment by virginiayung** [comment]
I've compared the output of rio-toa's [reflectance calculations](https://github.com/mapbox/rio-toa/blob/master/rio_toa/reflectance.py#L13) with results derived from [espa 's TOA ]formulas(https://github.com/USGS-EROS/espa-surface-reflectance/blob/master/not-validated-prototype-lasrc/c_version/src/l8sr_algorithm.txt#L89-L132). The two are identical. The only difference is in the threshold & rescaling. Espa sets a range for valid scaled reflectance value to be between [-2,000 and 16,000](https://github.com/USGS-EROS/espa-surface-reflectance/blob/master/not-validated-prototype-lasrc/c_version/src/l8sr_algorithm.txt#L31-L32) (equivalent to unscaled values [-0.2, 1.6]). We can consider adding this bounds as an option to be more consistent with scaling of the LaSRC's  TOA: https://github.com/mapbox/rio-toa/issues/12#issuecomment-234991325. At this point we can safely rule out the non-per-pixel sun elevation is 👍  I'm currently looking into the 512 band issue @celoyd found :point_up:https://github.com/mapbox/rio-toa/issues/15#issuecomment-234687777: 

#### Original `LC80460292016177LGN00_B3.TIF` vs. TOA (scene center sun elevation angle):

![screen shot 2016-07-26 at 4 06 09 pm](https://cloud.githubusercontent.com/assets/4450007/17158433/29c32a6c-534b-11e6-81d3-50ee8f05ea84.png)

#### TOA (scene center sun elevation angle) vs. per-pixel sun elevation TOA:

![screen shot 2016-07-26 at 4 06 17 pm](https://cloud.githubusercontent.com/assets/4450007/17158441/3724f4ce-534b-11e6-9b3c-d27da98d51b5.png)

**Comment by virginiayung** [comment]
##### Solution: Fixed in https://github.com/mapbox/rio-toa/pull/16

Summary: 
The anding issues we had observed☝️ were due to incorrect calculations of the latitude array. The original latitude outputs were flipped and were fixed by https://github.com/mapbox/rio-toa/pull/16/commits/eb6c435c81f16e7eb0c7ab88de2961d0d109ada3: 

from

```
 (lat.astype(np.float32) * yCell) + s + (yCell / 2.0))
```

to 

```
 (np.flipud(lat) * yCell) + s + (yCell / 2.0))
```

More documentation on [numpy.flipud](http://docs.scipy.org/doc/numpy/reference/generated/numpy.flipud.html)
